*: add snapshots

asubiotto opened the pull request:
This commit adds the ability to write a database snapshot at a given tx to disk
and load a snapshot into a database. This commit does not yet add a
public-facing API for snapshots.

The motivation for this change is to improve use cases where users enable the
WAL. Snapshots aim to solve two problems:
1) WAL replay can be very slow given that each insert needs to be replayed to
get to a "final" state. Snapshots can be loaded into a database in O(1) time
with respect to the number of inserts performed against the database instead
of the O(n) inserts required during a WAL replay.
2) With multiple tables, the WAL can only be truncated when all tables have
been persisted/rotated out and only at the minimum persisted tx across all
tables. If the user has a small table that doesn't change often and is
therefore not persisted/rotated out, the WAL can never be truncated. Snapshots
allow us to persist these small tables and truncate the WAL at the snapshot
txn.

Addresses #345 (will close it once this snapshot code is actually used)

metalmatze commented on snapshot.go:
```diff
@@ -0,0 +1,359 @@
+// segmentName returns a 20-byte textual representation of a snapshot file name
+// at a given txn used for lexical ordering.
+func snapshotFileName(tx uint64) string {
+	return fmt.Sprintf("%020d.fdbs", tx)
```
Couldn't we use ULIDs for this too? Not sure about the intent here.

asubiotto replied:
This encodes the transaction into the filename, similar to how the WAL code does it. Is there a good reason to use ULIDs instead of this?

metalmatze replied:
Not really, makes sense. Let's leave it as is.

metalmatze commented on snapshot.go:
```diff
@@ -0,0 +1,359 @@
+	}
+	defer f.Close()
+
+	if err := func() error {
+		if err := writeSnapshot(ctx, tx, db, f); err != nil {
+			return err
+		}
+		return f.Sync()
+	}(); err != nil {
```
Interesting, did you use a closure here to encapsulate the code so the context error wrapping can be done once below?

asubiotto replied:
Yes, it allows me to centralize cleanup code.

metalmatze reviewed (commented):
Really good stuff. Looking forward.

asubiotto commented on snapshot.go:
```diff
@@ -0,0 +1,359 @@
+		return 0, err
+	}
+
+	for i := len(files) - 1; i >= 0; i-- {
```
👍🏼 I'm planning on adding a high-level comment on how a snapshot is encoded

asubiotto commented on snapshot_test.go:
```diff
@@ -0,0 +1,152 @@
+	t.Run("WithConcurrentWrites", func(t *testing.T) {
+
+	})
+	// TODO(asubiotto): Maybe implement an Equal method on a table?
```
Yep, this is a leftover thought. If I do, it will only be a testing helper, but I still need to see if I need it for the tests.

thorfour commented:
Can you add to the description more info about this feature? It's unclear to me the trade-offs being made here.

asubiotto commented:
This PR is now ready for a review. I added a checksum and moved the version number outside the protobuf. I also added a more descriptive file-level comment as well as a commit message.

metalmatze reviewed (approved):
Very nice improvements!
Great comment on the overall snapshots and great tests!

thorfour reviewed (commented):
If we have a worse case insertion where every insert is the least we'd have a part per insert, which if I understand the LoadSnapshot function correctly that would be O(n) for loading a snapshot correct?

I mentioned in a another comment but I'd be really curious how it would perform to rewrite our WAL this way where we  write to the way the granule ID (or something) where a part is written, and then write the part to the WAL. So we could replay without having to perform any of the comparison logic that comes with inserts.

asubiotto commented:
> If we have a worse case insertion where every insert is the least we'd have a part per insert, which if I understand the LoadSnapshot function correctly that would be O(n) for loading a snapshot correct?

Yes, but that is a very very special case. You would have to have no compactions occur before the snapshot. Since snapshots would be taken periodically, you would not pay O(n) cost (where n = number of inserts) starting from the first compaction, and before that first compaction, n should be very small (we're talking a 100ms window of insertions, given that's our default compaction interval).

> I mentioned in a another comment but I'd be really curious how it would perform to rewrite our WAL this way where we write to the wal the granule ID (or something that identifies the granule) where a part is written, and then write the part to the WAL. So we could replay without having to perform any of the comparison logic that comes with inserts.

The problem is that as we perform more writes and granules are split etc... how do you maintain the idea of a granule ID? There's also the problem that if you're writing parts to the WAL, it's not really a write *ahead* log any more, since you're writing the result of writing the data to the database. Additionally, parts also change, so how do you maintain correctness during compactions? If we're not going over and rewriting WAL entries, we'd be appending these "state" changes but I'm not sure if it makes sense to do that.

asubiotto commented on snapshot.go:
```diff
@@ -0,0 +1,479 @@
+						if err != nil {
+							return err
+						}
+						resultParts = append(resultParts, parts.NewPart(partMeta.Tx, serBuf, partOptions))
```
Responded in the top-level PR comment.

asubiotto commented on snapshot.go:
```diff
@@ -0,0 +1,479 @@
+package frostdb
```
Agreed. I was looking at doing this, but I think came to the conclusion that we needed access to some db internals that would make the API a bit murky. Can take another look at doing this.

asubiotto replied:
I tried creating the snapshots package and a lot of internal dependencies can be exported, however when loading a snapshots, the code manipulates a table's active block index. It feels wrong to export this behavior and moving just the index code to the `frostdb` package doesn't seem like a clean break. What do you think?

asubiotto replied:
I can maybe try to return the new index from the snapshot code to be loaded. Not sure I like it but will work on a commit so that we can see the final result.

asubiotto replied:
Hmm, but since we can be loading multiple tables we'd have to return some way to map table names to indexes. I think it's probably not worth doing so at this point...

thorfour commented:
> 

Totally fair,  I was mostly riffing an idea that popped into my head when I reviewed. Was thinking each granule would be assigned an immutable ID upon creation and then each write into the granule would make it into the WAL `<id><data>`.
I think it would still constitute a write ahead log as we wouldn't return success of the write until it was completed but I take your point and my random thought definitely wasn't well thought out.

asubiotto commented:
Gotcha. No problem! I think it's great to talk about alternatives.
